### DISCUSSION TITLE
Supporting multiple database vendors via env variables

### DISCUSSION THREAD

**Discussion by dasteg** [quarkusio/quarkus#39700] [Community]
Hello,

i am working on a quarkus application that should support mssql and postgresql. since the db-kind property is a build-time property i can't simply use an env variable to switch between database vendors.

currently we are using the approach to have two executions in the quarkus maven plugin and basically creating the quarkus-app twice. this leads also to two containers with -mssql and -postgresql suffix.

```xml
<plugin>
	<groupId>io.quarkus</groupId>
	<artifactId>quarkus-maven-plugin</artifactId>
	<extensions>true</extensions>
	<executions>
		<execution>
			<id>mssql</id>
			<goals>
				<goal>build</goal>
				<goal>generate-code</goal>
				<goal>generate-code-tests</goal>
			</goals>
			<configuration>
				<properties>
					<databaseType>mssql</databaseType>
					<quarkus.profile>database-mssql</quarkus.profile>
					<quarkus.package.output-directory>quarkus-app-mssql</quarkus.package.output-directory>
					<quarkus.package.filter-optional-dependencies>true</quarkus.package.filter-optional-dependencies>
					<quarkus.package.included-optional-dependencies>org.flywaydb:flyway-sqlserver::jar,io.quarkus:quarkus-jdbc-mssql::jar</quarkus.package.included-optional-dependencies>
					<quarkus.container-image.tag>${project.version}-mssql</quarkus.container-image.tag>
				</properties>
			</configuration>
		</execution>
		<execution>
			<id>postgresql</id>
			<goals>
				<goal>build</goal>
				<goal>generate-code</goal>
				<goal>generate-code-tests</goal>
			</goals>
			<configuration>
				<properties>
					<databaseType>postgresql</databaseType>
					<quarkus.profile>database-postgresql</quarkus.profile>
					<quarkus.package.runner-suffix>-runner-postgresql</quarkus.package.runner-suffix>
					<quarkus.package.output-directory>quarkus-app-postgresql</quarkus.package.output-directory>
					<quarkus.package.filter-optional-dependencies>true</quarkus.package.filter-optional-dependencies>
					<quarkus.package.included-optional-dependencies>io.quarkus:quarkus-jdbc-postgresql::jar</quarkus.package.included-optional-dependencies>
					<quarkus.container-image.tag>${project.version}-postgresql</quarkus.container-image.tag>
				</properties>
			</configuration>
		</execution>
	</executions>
</plugin>
```
i am not really happy with this solution because i'd actually prefer to ultimately get some all-in-one solution where an env variable defines which database should be used. one idea that came to my mind was to put both quarkus-app-* folders into the container and let the entrypoint decide which app to start. despite the container size that would be the preferable solution.

are there any other ways to do this ? i didn't find anything and wonder if i am the first person to have this issue/requirement.

looking forward to any input :)

**Comment by gsmet** [comment]
Yes, it is a build time property because it actually affects quite a lot of things.

It's hard to know exactly what you could do without having all the info but I have seen people:
- have both JDBC drivers around, and enabling/disabling named datasources depending on an env var
- apps like Keycloak use reaugmentation (see https://quarkus.io/guides/reaugmentation) but it's a bit complex to put in place and I'm not sure it will actually solve your problem

Anyway, I would recommend to have a look at the two approaches and if they are feasible for you and solve your problem.

**Reply by dasteg** [reply]
hi,

that actually helped a lot. reaugmentation indeed seems to be too complex for my use case and also would require some changes in my pipelines .

but the enabling/disabling of datasources seems to be exactly what i was looking for. using a custom entrypoint i can simply decide which quarkus profile to start and based on that configure the respective datasources. now that i know that this is an option the documentation for datasources also makes more sense.
